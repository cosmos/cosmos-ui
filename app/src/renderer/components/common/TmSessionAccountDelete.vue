<template>
  <div class="tm-session">
    <tm-form-struct :submit="onSubmit" class="tm-session-container">
      <div class="tm-session-header">
        <a
          @click="setState('sign-in')"
        >
          <i class="material-icons">arrow_back</i>
        </a>
        <div class="tm-session-title">
          Remove Account
        </div>
        <a @click="help"><i class="material-icons">help_outline</i></a>
      </div>
      <div class="tm-session-main">
        <tm-form-group
          :error="$v.deletionPassword.$error"
          field-id="sign-in-password"
          field-label="Password"
        >
          <tm-field
            id="sign-in-password"
            v-model="deletionPassword"
            type="password"
            placeholder="Enter your password"
          />
          <tm-form-msg
            v-if="!$v.deletionPassword.required"
            name="Password"
            type="required"
          />
          <tm-form-msg
            v-if="!$v.deletionPassword.minLength"
            name="Password"
            type="minLength"
            min="10"
          />
        </tm-form-group>
        <tm-form-group
          :error="$v.deletionWarning.$error"
          field-id="sign-up-warning"
          field-label=" "
        >
          <div class="tm-field-checkbox">
            <div class="tm-field-checkbox-input">
              <input
                id="sign-up-warning"
                v-model="deletionWarning"
                type="checkbox"
              >
            </div>
            <label
              class="tm-field-checkbox-label"
              for="sign-up-warning"
            >
              I understand that Cosmos cannot recover deleted accounts without
              the passphrase.
            </label>
          </div>
          <tm-form-msg
            v-if="!$v.deletionWarning.required"
            name="Deletion confirmation"
            type="required"
          />
        </tm-form-group>
      </div>
      <div class="tm-session-footer">
        <tm-btn
          icon="exit_to_app"
          value="Sign Out and Remove Account"
          size="lg"
        />
      </div>
    </tm-form-struct>
  </div>
</template>

<script>
import { required, minLength, sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
export default {
  name: `tm-session-account-delete`,
  components: {
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct
  },
  data: () => ({ 
    deletionPassword: ``,
    deletionWarning: false 
  }),
  mounted() {
    this.$el.querySelector(`#sign-in-password`).focus()
  },
  methods: {
    help() {
      this.$store.commit(`setModalHelp`, true)
    },
    setState(value) {
      this.$store.commit(`setSessionModalView`, value)
    },
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        const success = await this.$store.dispatch(`deleteKey`, {
          password: this.deletionPassword
        })
        if (success) {
          this.setState(`welcome`)
          this.$store.commit(`notify`, {
            title: `Account Deleted`,
            body: `You have successfully deleted the account 'default'`
          })
        }
      } catch (error) {
        this.$store.commit(`notifyError`, {
          title: `Account Deletion Failed`,
          body: error.message
        })
      }
    }
  },
  validations: () => ({
    deletionPassword: { required, minLength: minLength(10) },
    deletionWarning: { required: sameAs(() => true) }
  })
}
</script>
