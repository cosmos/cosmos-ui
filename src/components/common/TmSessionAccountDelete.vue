<template>
  <div class="session">
    <TmFormStruct :submit="onSubmit" class="session-container">
      <div class="session-header">
        <a @click="setState('sign-in')">
          <i class="material-icons">arrow_back</i>
        </a>
        <div class="session-title">
          Remove Account
        </div>
      </div>
      <div class="session-main">
        <TmFormGroup
          :error="$v.deletionPassword.$error"
          field-id="sign-in-password"
          field-label="Password"
        >
          <TmField
            id="sign-in-password"
            v-model="deletionPassword"
            type="password"
            placeholder="Enter your password"
          />
          <TmFormMsg
            v-if="!$v.deletionPassword.required"
            name="Password"
            type="required"
          />
          <TmFormMsg
            v-if="!$v.deletionPassword.minLength"
            name="Password"
            type="minLength"
            min="10"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.deletionWarning.$error"
          field-id="sign-up-warning"
          field-label=" "
        >
          <div class="field-checkbox">
            <div class="field-checkbox-input">
              <input
                id="sign-up-warning"
                v-model="deletionWarning"
                type="checkbox"
              />
            </div>
            <label class="field-checkbox-label" for="sign-up-warning">
              I understand that Cosmos cannot recover deleted accounts without
              the passphrase.
            </label>
          </div>
          <TmFormMsg
            v-if="!$v.deletionWarning.required"
            name="Deletion confirmation"
            type="required"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn icon="exit_to_app" value="Sign Out and Remove Account" />
      </div>
    </TmFormStruct>
  </div>
</template>

<script>
import { required, minLength, sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
export default {
  name: `session-account-delete`,
  components: {
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct
  },
  data: () => ({
    deletionPassword: ``,
    deletionWarning: false
  }),
  mounted() {
    this.$el.querySelector(`#sign-in-password`).focus()
  },
  methods: {
    setState(value) {
      this.$store.commit(`setSessionModalView`, value)
    },
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        const success = await this.$store.dispatch(`deleteKey`, {
          password: this.deletionPassword
        })
        if (success) {
          this.setState(`welcome`)
          this.$store.commit(`notify`, {
            title: `Account Deleted`,
            body: `You have successfully deleted the account 'default'`
          })
        }
      } catch (error) {
        this.$store.commit(`notifyError`, {
          title: `Account Deletion Failed`,
          body: error.message
        })
      }
    }
  },
  validations: () => ({
    deletionPassword: { required, minLength: minLength(10) },
    deletionWarning: { required: sameAs(() => true) }
  })
}
</script>
