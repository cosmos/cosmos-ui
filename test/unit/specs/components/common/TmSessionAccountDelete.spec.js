import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionAccountDelete from "common/TmSessionAccountDelete"

describe(`TmSessionAccountDelete`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)

  let wrapper, $store

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(async () => true),
    }

    wrapper = shallowMount(TmSessionAccountDelete, {
      localVue,
      mocks: {
        $store
      }
    })
    wrapper.setData({
      deletionPassword: ``,
      deletionWarning: false
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should go back to the login screen on click`, () => {
    wrapper
      .findAll(`.tm-session-header a`)
      .at(0)
      .trigger(`click`)
    expect($store.commit.mock.calls[0]).toEqual([
      `setSessionModalView`,
      `sign-in`
    ])
  })

  it(`should go back on successful deletion`, async () => {
    wrapper.setData({
      deletionPassword: `1234567890`,
      deletionWarning: true
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toEqual([
      `setSessionModalView`,
      `welcome`
    ])
  })

  it(`should show error if password not 10 long`, async () => {
    wrapper.setData({
      deletionPassword: `123`,
      deletionWarning: true
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show error if deletionWarning is not acknowledged`, async () => {
    wrapper.setData({
      deletionPassword: `1234567890`,
      deletionWarning: false
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show a notification if deletion failed`, async () => {
    const error = {
      message: `Error body`
    }

    $store.commit = jest.fn()
    $store.dispatch = jest.fn().mockRejectedValue(error)

    wrapper.setData({
      deletionPassword: `1234567890`,
      deletionWarning: true
    })

    await wrapper.vm.onSubmit()
    expect($store.commit).toHaveBeenCalledWith(`notifyError`, { body: `Error body`, title: `Account Deletion Failed` })
  })
})
