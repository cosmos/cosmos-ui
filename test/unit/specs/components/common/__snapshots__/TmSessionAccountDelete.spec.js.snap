// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionAccountDelete has the expected html structure 1`] = `
<div
  class="tm-session"
>
  <tm-form-struct-stub
    class="tm-session-container"
    submit="function () { [native code] }"
    width=""
  >
    <div
      class="tm-session-header"
    >
      <a>
        <i
          class="material-icons"
        >
          arrow_back
        </i>
      </a>
       
      <div
        class="tm-session-title"
      >
        
        Remove Account
      
      </div>
       
      <a>
        <i
          class="material-icons"
        >
          help_outline
        </i>
      </a>
    </div>
     
    <div
      class="tm-session-main"
    >
      <tm-form-group-stub
        fieldid="sign-in-password"
        fieldlabel="Password"
      >
        <tm-field-stub
          id="sign-in-password"
          placeholder="Enter your password"
          type="password"
          value=""
        />
         
        <tm-form-msg-stub
          msg=""
          name="Password"
          type="required"
        />
         
        <!---->
      </tm-form-group-stub>
       
      <tm-form-group-stub
        fieldid="sign-up-warning"
        fieldlabel=" "
      >
        <div
          class="tm-field-checkbox"
        >
          <div
            class="tm-field-checkbox-input"
          >
            <input
              id="sign-up-warning"
              type="checkbox"
            />
          </div>
           
          <label
            class="tm-field-checkbox-label"
            for="sign-up-warning"
          >
            
            I understand that Cosmos cannot recover deleted accounts without
            the passphrase.
          
          </label>
        </div>
         
        <tm-form-msg-stub
          msg=""
          name="Deletion confirmation"
          type="required"
        />
      </tm-form-group-stub>
    </div>
     
    <div
      class="tm-session-footer"
    >
      <tm-btn-stub
        icon="exit_to_app"
        size="lg"
        value="Sign Out and Remove Account"
      />
    </div>
  </tm-form-struct-stub>
</div>
`;
